Step3 진행하여 리뷰 요청 드립니다.

deokmoon opened the pull request:
안녕하세요, 강덕문입니다.

요구사항을 주체 - 행위로 묶어 객체를 설계하려고 노력하였고,
도메인에 대한 TDD를 진행하는데는 수월하였습니다.
하지만 도메인을 모두 설계하고, 각각의 도메인을 수행하는 로직을 구현할 때 많은 어려움을 느꼈습니다.

'UI분리를 해야하는데, 이렇게 하는게 맞는 것일까?' 라는 고민을 가지며 진행하였습니다.
요구사항에 맞는 설계인지 자신은 없지만, 리뷰 요청드립니다.

질문사항 추가 작성합니다!
일급콜렉션을 적용 중 어려움을 겪어, 민제리뷰어님의 조언을 얻고자 질문드립니다.
처음 생각 했던 내용은 아래와 같습니다.
 - LottoNumbers: 로또 숫자범위의 숫자들(6개)을 가짐 -> List<Integer>
 - Lotto: 하나의 LottoNumbers 를 가짐
 - Lottos: 여러개의 Lotto를 가짐
 - Buyer: Lottos 를 구매함
위와 같이 초기 설계를 했지만, Buyer로 부터 Lotto 번호를 얻으려면, 중복된? 메서드, 메서드를 타고타고 가는 느낌이 있었습니다.
그래서 최종적으론 Lottos를 삭제하고 Buyer에 List<Lotto>로 정의하여 사용했습니다.

뭔가 일급콜렉션을 제대로 활용하지 못하는 것 같아, 처음 설계가 잘못된 것인지... 하면서 혼란이 왔습니다.
혹시 해당 미션에 민제리뷰어님이라면 일급콜렉션을 어디에 적용하셨을 것 같은지 여쭤보고 싶었습니다.

오늘도 잘 부탁드립니다. 😊 
감사합니다.

deokmoon committed:
Test: step2 피드백반영 ParmeterizedTest 활용

deokmoon committed:
Refactor: step2 피드백반영 Pattern 상수화, 상수와 메서드 사이 개행 추가

deokmoon committed:
Docs: Step3 요구사항 정리

deokmoon committed:
Test: Money(돈) 원시값 포장 클래스 테스트코드 작성
 - 입력받은 금액 String to Int 변환 저장 테스트
 - 금액 문자열 존재 시 예외 테스트

deokmoon committed:
Test: Money(돈) 원시값 포장 클래스 테스트코드 파라미터수정
 - 예외 파라미터 중 "" 추가 적용

deokmoon committed:
Feat: Money(돈) 원시값 포장 클래스 프로덕트코드 작성
 - 금액 문자열 검증
 - 입력받은 금액 String to Int 변환 저장

deokmoon committed:
Test: 6개의 로또 번호 갖는 일급콜렉션 테스트코드 작성
 - 6개의 숫자를 갖는지 확인

deokmoon committed:
Feat: 6개의 로또 번호 갖는 일급콜렉션 프로덕트코드 작성

deokmoon committed:
Test: 로또 객체의 테스트코드 작성
 - 로또 클래스는 기본 가격과 6개의 숫자를 가짐
 - 6개의 숫자는 일급콜렉션에 책임을 일임

deokmoon committed:
Test: 로또 객체의 테스트코드 수정
 - 가격을 int가 아닌 원시값 포장한 Money 사용

deokmoon committed:
Feat: 로또 객체 설계 및 구현
 - Money 형식의 기본 가격을 가짐
 - 일급콜렉션 LottoNumbers를 통해 6개의 숫자를 가짐

deokmoon committed:
Refactor: 잘못된 Import 제거

deokmoon committed:
Docs: Lottos 일급콜렉션 관련 내용 추가 작성

deokmoon committed:
Test: Lottos 일급콜렉션 테스트 코드 작성

deokmoon committed:
Feat: Lottos 일급콜렉션 프로덕트 코드 구현

deokmoon committed:
Feat: 로또 기본금액 상수화

deokmoon committed:
Refactor: 로또 기본금액 상수화에 따른 테스트코드 수정

deokmoon committed:
Test: Money 원시값 포장 클래스 금액차감 기능 테스트코드 작성

deokmoon committed:
Feat: Money 원시값 포장 클래스 금액차감 기능 구현

deokmoon committed:
Docs: Lottos 삭제관련 내용 작성
 - 일급콜렉션 의도로 Lottos를 작성했으나 중복코드인 것 같아 삭제

deokmoon committed:
Refactor: Lottos 관련 프로덕트, 테스트코드 삭제
 - 일급콜렉션 의도로 Lottos를 작성했으나 중복코드인 것 같아 삭제

deokmoon committed:
Refactor: 차감 기능 파라미터 String으로 변경(테스트코드, 프로덕트코드 일괄 적용)

deokmoon committed:
Fix: Lottos 삭제로 인한 버그 수정
- LottsTest가 아닌 LottoTest를 삭제 후 커밋되어, LottosTest를 LottoTest 리네임 후 LottoTest의 테스트 코드로 재작성

deokmoon committed:
Test: 구매자 객체 테스트 코드 작성
 - 구매자 객체 잔액 저장 테스트
 - 구매자 객체 로또 구입 기능 테스트

deokmoon committed:
Feat: 구매자 객체 구현
 - 구매자 객체 잔액 저장 기능 구현
 - 구매자 객체 로또 구입 기능 구현

deokmoon committed:
Feat: String to Int 변환 공통 Util 생성
 - 중복코드 발생으로 공통 메서드로 처리

deokmoon committed:
Refactor: String to Int 변환 공통 메서드화로 인한 수정

deokmoon committed:
Test: 문자열 숫자 변환 공통메서드화로 인한 테스트코드 재작성

deokmoon committed:
Test: 입력받은 문자열로 당첨로또 저장 테스트코드 작성

deokmoon committed:
Feat: 입력받은 문자열로 당첨로또 저장 기능 구현 및 구분자 상수화

deokmoon committed:
Docs: 기능 요구사항 정리한 내용 객체와 매핑

deokmoon committed:
Test: 로또번호 수동입력 시 예외처리 테스트코드 추가 작성

deokmoon committed:
Feat: 로또번호 수동입력 시 예외처리 추가 구현

deokmoon committed:
Feat: 로또번호 수동입력 시 참조 로또 디폴트 사이즈 정의

deokmoon committed:
Refactor: 불필요한 Import 삭제

deokmoon committed:
Feat: Input Print 구매개수 관련 기능 추가

deokmoon committed:
Refactor: 사용자 단에서 로또 번호 가져올 시 중복코드 발생 제거
 - Lotto.getLottoNumbers().getLottoNumbers();
  -> Lotto.getLottoNumbers(); 되도록 수정

deokmoon committed:
Feat: 객체 생성 및 로또 일치 프로세스 구현

deokmoon committed:
Feat: 당첨 통계를 위한 일치개수, 금액 Enum클래스 구현

deokmoon committed:
Feat: 최소 당첨 숫자개수 상수 추가

deokmoon committed:
Feat: 당첨번호, 당첨금액 기준으로 가져오도록 기능 구현

deokmoon committed:
Feat: 당첨여부 판단하는 기능 구현

deokmoon committed:
Feat: InputView, ResultView 구현

deokmoon committed:
Feat: 로또 프로그램을 시작하는 Application 구현

minzzang commented on README.md:
```diff
@@ -20,4 +20,58 @@
+* TDD 구현하여 단위 테스트 작성(UI 로직 제외)
+* 핵심 로직을 구현하는 코드와 UI 담당하는 로직을 구분
+* indent는 1까지 허용
+* 메서드의 길이는 15라인까지 허용
```
요구사항 정리 잘하셨네요~ 👍

minzzang commented on src/main/java/lotto/auto/Lotto.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.List;
+
+public class Lotto {
+    private Money price;
```
로또에서 Money 변수를 가지고 있는 이유를 알 수 있을까요~? 상수로 선언해도 괜찮을 것 같은데 덕문님 생각이 궁금하네요 😄

minzzang commented on src/main/java/lotto/auto/LottoNumbers.java:
```diff
@@ -0,0 +1,58 @@
+        Collections.sort(lottoNumbers);
+    }
+
+    private void makeLottoNumbers() {
```
로또 번호 생성에 대한 역할을 다른 클래스로 위임해보는건 어떨까요~?

minzzang commented on src/main/java/lotto/auto/DestinyMatch.java:
```diff
@@ -0,0 +1,56 @@
+
+    private int isMatched(int num, LottoNumbers winningNumbers) {
+        if (winningNumbers.getLottoNumbers().contains(num)) {
+            return 1;
```
리터럴 값을 의미있는 변수로 포장해 보는건 어떨까요~?
https://limdingdong.tistory.com/5

minzzang commented on src/main/java/lotto/auto/Buyer.java:
```diff
@@ -0,0 +1,34 @@
+            this.amount.substractMoney(DEFAULT_PRICE);
+            cnt++;
+        }
+        System.out.println(cnt + "개를 구매했습니다.");
```
도메인 로직에서 view 영역을 의존하는 것 같네요. cnt를 반환하고 view에서 처리하는 건 어떤가요~?

minzzang commented on src/main/java/lotto/auto/DestinyMatch.java:
```diff
@@ -0,0 +1,56 @@
+        resultView.printResult(inputMoney, matchWinningNumbers(buyer.getLotto(), winningNumbers));
+    }
+
+    private Map<Integer, Integer> matchWinningNumbers(List<Lotto> lottoList, LottoNumbers winningNumbers) {
```
당첨 통계를 계산하는 로직은 누가 담당하면 좋을까요? Map을 사용해서 계산을 하고 있는데 이 부분도 일급 컬렉션으로 만드는건 어떨까요~?

minzzang commented on src/main/java/lotto/auto/common/LottoAutoUtils.java:
```diff
@@ -0,0 +1,14 @@
+public class LottoAutoUtils {
+    public int stringToNumber(String number) {
+        if (number == null) {
+            throw new IllegalArgumentException();
```
의미 있는 예외 메시지를 작성해보는 건 어떨까요~?

minzzang commented on src/main/java/lotto/auto/ui/ResultView.java:
```diff
@@ -0,0 +1,47 @@
+
+public class ResultView {
+
+    public ResultView() {
```
기본 생성자 이외에 다른 생성자가 없는데 작성하신 이유가 있을까요~?

minzzang commented on src/main/java/lotto/auto/Buyer.java:
```diff
@@ -0,0 +1,34 @@
+
+import static lotto.auto.common.Constants.DEFAULT_PRICE;
+
+public class Buyer {
```
질문에 대한 저의 생각을 말씀드리자면 Lottos -> Lotto -> LottoNumbers로 타고 들어가도 괜찮다고 생각을 합니다. 다만 Lottos가 LottoNubmers에 값을 가져올 필요가 있을 경우 Lotto에게 메시지를 전달해 가져오는 방식으로 구현하시면 될 것 같아요. 즉 Lottos는 Lotto의 내부 구현(LottoNumbers)을 알지 못하게 하면 될 것 같습니다.

일급 컬렉션을 적용해야 한다면 저도 마찬가지로 Lottos, Lotto에 적용을 하고 컬렉션을 사용하는 부분에서는 최대한 일급 컬렉션을 적용해보려고 할 것 같습니다.

설계 하시기 어려울 때는 손으로 직접 클래스도 그려보고 어떤 메시지를 주고 받을지도 그려보면 조금 더 정리가 잘 될 거에요. 또한 도메인에 대해 이해하기 어려울 때 TDD 방식으로 바로 테스트 코드를 먼저 작성하면 코드를 작성하기 더 어려울 수도 있으니 먼저 코드를 쭉 작성하고 테스트 코드를 짜는 방법도 괜찮은 방법 같습니다. 완벽한 설계를 한번에 하고 코드를 작성하면 물론 좋겠지만 그러기 어렵기 때문에 설계하고 코드 짜보고 수정하고 다시 설계하고 반복하다보면 더 좋은 설계가 나올 수 있을 거에요.

혹시 아직 읽어보시지 않았다면 조영호 님의 객체지향의 사실과 오해 이 책을 추천드립니다. 그리고 일급 컬렉션은 향로님 글을 보면 이해가 조금 잘 되실거에요.
https://jojoldu.tistory.com/412

minzzang commented on src/main/java/lotto/auto/LottoNumbers.java:
```diff
@@ -0,0 +1,58 @@
+    private final int BEGIN_INDEX = 0;
+    private final int LOTTO_NUMBER_LIST_SIZE = 6;
+
+    private List<Integer> lottoNumbers;
```
로또 번호도 포장을 하면 번호에 대한 validation처리도 위임할 수 있고 캐싱을 적용할 수도 있을 것 같아요 😄

minzzang reviewed (approved):
안녕하세요 덕문님~!

3단계 미션 잘 수행해주셨습니다. !! :+1:
도메인을 분리하는데 많은 고민과 노력을 한 흔적이 보이네요 :+1:

다만, 아직 책임을 조금 더 분리할 수 있을 거 같습니다!
질문에 대한 답변과 코멘트 달았는데 확인 부탁드려요 ~ :pray:

이해가 안되는 부분이 있거나 고민이 있으시면 편하게 DM 해주세요 :smile:

minzzang reviewed (changes requested):
안녕하세요 덕문님~!

3단계 미션 잘 수행해주셨습니다. !! :+1:
도메인을 분리하는데 많은 고민과 노력을 한 흔적이 보이네요 :+1:

다만, 아직 책임을 조금 더 분리할 수 있을 거 같습니다!
질문에 대한 답변과 코멘트 달았는데 확인 부탁드려요 ~ :pray:

이해가 안되는 부분이 있거나 고민이 있으시면 편하게 DM 해주세요 :smile:

deokmoon committed:
Test: 로또, 일급콜렉션 LotttoNumbers 테스트 코드 수정
 - 리뷰 피드백 반영으로, 클랙스 역할 책임에 따른 테스트 코드 수정
 - LotttoNumbers 일급콜렉션으로 Validation 책임 추가

deokmoon committed:
Refactor: Lotto 클래스, 로또 번호 생성 책임 부여
 - 리뷰 피드백 반영으로, 클랙스 역할 책임 조정

deokmoon committed:
Refactor: Lotto 번호 일급콜렉션으로 Refactoring
 - 번호 생성에 대한 책임 제외
 - 번호 validation 추가

deokmoon committed:
Refactor: Lottos 일급콜렉션 추가, List<Lotto> -> Lottos 수정

deokmoon committed:
Refactor: Lottos 일급콜렉션 추가, List<Lotto> -> Lottos 수정

deokmoon committed:
Refactor: 공통 상수로 쓰일 것으로 예상되는 값 상수클래스에 적용

deokmoon committed:
Fix: 생성자 버그 수정

deokmoon committed:
Feat: 매칭 결과 저장하는 일급콜렉션 클래스 구현

deokmoon committed:
Feat: 매칭 결과 책임을 수행하는 클래스 생성

deokmoon committed:
Refactor: 매칭 결과 책임 분리

deokmoon committed:
Refactor: 일급콜렉션 Rank 활용으로 리팩토링

deokmoon committed:
Refactor: 책임 분리 및 Refactoring

deokmoon committed:
Feat: 공통상수 추가

deokmoon committed:
Refactor: 상수 추가 리팩토링

deokmoon committed:
Refactor: 예외 메세지 작성

deokmoon committed:
Fix: 출력 메세지 조정을 위한 버그 수정
 - 상수 순서 수정
 - 메세지 버그 수정

deokmoon committed:
Refactor: Rank Null처리 추가 적용

deokmoon committed:
Feat: 테스트코드 작성을 위한 생성자 추가

deokmoon committed:
Feat: 테스트코드 작성을 위한 생성자 추가

deokmoon committed:
Fix: 파라미터 타입 수정

deokmoon committed:
Refactor: 소수점 고통 상수화 적용

deokmoon committed:
Refactor: 이넘명 수정

deokmoon committed:
Refactor: 숫자 매칭 프로세스 테스트 코드 작성

deokmoon committed:
Refactor: 로또 꽝인 경우 예외처리 로직 적용

deokmoon committed:
Refactor: 로또 꽝인 경우 예외처리 로직 적용에 따른 상수추가

minzzang commented on src/main/java/lotto/auto/LottoNumbers.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public List<Integer> getNumbers() {
+        return numbers;
```
방어적 복사나 불변 객체를 반환하면 더 좋을 것 같아요!
https://tecoble.techcourse.co.kr/post/2021-04-26-defensive-copy-vs-unmodifiable/

minzzang reviewed (approved):
안녕하세요 덕문님~!

피드백 잘 반영해주셨네요!!
고생하셨습니다. :pray:
